### PR TITLE
Clean up getindex

### PR DESCRIPTION
**Pull request by bkamins**
This is a first (simpler) part to make #1380 clean after Julia 1.0. It concentrates on `getindex`.
What was deprecated is now an error.
Additionally I have reviewed whole DataFrames.jl to fix all `getindex` occurrences. There are some new deprecation warnings.

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -368,7 +368,17 @@ mutable struct StackedVector <: AbstractVector{Any}
     components::Vector{Any}
 end
 
+function Base.getindex(v::StackedVector,i::Bool)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Put these in deprecated.jl?

File: src/abstractdataframe/reshape.jl [hunk 2]
```diff
@@ -382,14 +392,19 @@ function Base.getindex(v::StackedVector,i::Real)
     v.components[j][k]
 end
 
-function Base.getindex(v::StackedVector,i::AbstractVector{I}) where I<:Real
+function Base.getindex(v::StackedVector,i::AbstractVector)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Shouldn't this be restricted to `<:Integer`?

File: src/abstractdataframe/reshape.jl [hunk 3]
```diff
@@ -437,17 +452,28 @@ mutable struct RepeatedVector{T} <: AbstractVector{T}
...
-function Base.getindex(v::RepeatedVector{T},i::Real) where T
+
+function Base.getindex(v::RepeatedVector, i::Integer)
     N = length(v.parent)
     idx = Base.fld1(mod1(i,v.inner*N),v.inner)
     v.parent[idx]
 end
-Base.getindex(v::RepeatedVector,i::AbstractRange) = getindex(v, [i;])
+
+Base.getindex(v::RepeatedVector,i::AbstractVector) = [v[j] for j in i]
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: I'm surprised there is no fallback doing the same thing already.

File: src/abstractdataframe/reshape.jl [hunk 4]
```diff
@@ -382,14 +392,19 @@ function Base.getindex(v::StackedVector,i::Real)
...
     result = similar(v.components[1], length(i))
     for idx in 1:length(i)
         result[idx] = v[i[idx]]
     end
     result
 end
 
+function Base.getindex(v::StackedVector, i::AbstractVector{Bool})
+    length(v) == length(i) || throw(BoundsError(v, idx))
+    v[findall(i)]
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Would `[v[x] for x in i if x]` be fast enough? It would avoid an allocation.

I'm also surprised there is no fallback doing the same thing already. Same for `getindex(v::RepeatedVector,i::AbstractVector)`.

File: src/dataframe/dataframe.jl [hunk 5]
```diff
@@ -302,7 +302,10 @@ end
 
 # df[:, SingleColumnIndex] => AbstractVector
 # df[:, MultiColumnIndex] => DataFrame
-Base.getindex(df::DataFrame, row_ind::Colon, col_inds) = df[col_inds]
+function Base.getindex(df::DataFrame, row_ind::Colon, col_inds)
+    Base.depwarn("indexing with colon as row will create a copy of rows in the future", :getindex)
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: "a copy" (not "of rows" more than "of columns"). Also point people at `df[col_inds]` to avoid the deprecation warning until we remove it.

File: src/dataframerow/dataframerow.jl [hunk 6]
```diff
@@ -17,7 +17,7 @@ function Base.getindex(r::DataFrameRow, idx::AbstractArray)
     return DataFrameRow(parent(r)[idx], row(r))
 end
 
-function Base.getindex(r::DataFrameRow, idx::Any)
+function Base.getindex(r::DataFrameRow, idx::ColumnIndex)
```
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl]: Does this need a deprecation?

File: src/other/index.jl [hunk 7]
```diff
@@ -134,18 +135,14 @@ end
...
     end
     getindex(x, collect(Missings.replace(idx, false)))
 end
 
 function Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Integer})
     if any(v -> v isa Bool, idx)
-        # TODO: this line should be changed to throw an error after deprecation
-        # throw(ArgumentError("Bool values except for Vector{Bool} are not allowed for column indexing"))
-        Base.depwarn("Indexing with Bool values is deprecated except for Vector{Bool}")
+        throw(ArgumentError("Bool values except for Vector{Bool} are not allowed for column indexing"))
```
Comment (nalimilan) [inline src/other/index.jl]: Or `AbstractVector{Bool}`?

File: test/data.jl [hunk 8]
```diff
@@ -190,8 +190,17 @@ module TestData
         @test size(dx) == (0, 3)
         @test names(dx) == [:variable, :value, :a]
 
-        stackdf(d1, :a)
         d1s = stackdf(d1, [:a, :b])
+        # TODO: add tests for Bool and Real for d1s columns 1 and 2 after deprecation of getindex
```
Comment (nalimilan) [inline test/data.jl]: Can you add these with `@test_broken`? That way we'll notice the failure when we remove the deprecations.

File: test/data.jl [hunk 9]
```diff
@@ -190,8 +190,17 @@ module TestData
...
+        # d1s[1][true]
+        # d1s[1][1.0]
+        # d1s[2][true]
+        # d1s[2][1.0]
+        
+        # Those two tests check indexing by Vector{Bool} that were broken earlier
```
Comment (nalimilan) [inline test/data.jl]: That's not a `Vector{Bool}`. Also "earlier" won't mean anything to future readers. :-)

File: src/abstractdataframe/reshape.jl (hunk 3 above)
Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: Good point. Should not be necessary.

File: src/abstractdataframe/reshape.jl (hunk 4 above)
Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: OK - let us try removing them.

File: src/dataframe/dataframe.jl (hunk 5 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: fixed

File: src/dataframerow/dataframerow.jl (hunk 6 above)
Comment (bkamins) [reply src/dataframerow/dataframerow.jl]: The only case offending was when `idx` was a colon, where a `DataFrame` was returned. I have fixed it (but we return `DataFrameRow` now as I guess this is more consistent, and the old behavior was unintended). We can add deprecation for it if you prefer.

File: src/other/index.jl (hunk 7 above)
Comment (bkamins) [reply src/other/index.jl]: fixed

File: test/data.jl (hunk 8 above)
Comment (bkamins) [reply test/data.jl]: added

File: test/data.jl (hunk 9 above)
Comment (bkamins) [reply test/data.jl]: fixed

**Comment by bkamins** [comment]
@nalimilan: I will add more tests to be sure that `getindex` is 100% clean in all corner cases and let you know when all tests pass. Then we can merge this I guess.

File: test/data.jl [hunk 10]
```diff
@@ -190,8 +190,24 @@ module TestData
         @test size(dx) == (0, 3)
         @test names(dx) == [:variable, :value, :a]
 
-        stackdf(d1, :a)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/data.jl]: fixed
